### PR TITLE
Object-tracking - Data preprocessing/transforms

### PR DESCRIPTION
**Pull request by ss-sinha**
# Peer Review Information

PR for SiamMask (Object Tracking Model) -1:

1. transforms.py: Added some transforms specific to object tracking
2. coco.py: Made a new class which encapsulates all pre-processing, download, etc.
3. Added entry for SiamMask in __init__
TODO: Tests 
# Pull Request Check List

- [ ] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

### CODE REVIEW

File: datasetinsights/io/transforms.py [hunk 1]
```diff
@@ -59,3 +61,144 @@ def __call__(self, sample):
         if self.target_size != -1:
             target = F.resize(target, self.target_size)
         return (img, target)
+
+
+class BlurImage:
```
Comment (adason) [inline datasetinsights/io/transforms.py:66]: Can you write unittests for each of these tansforms?

File: datasetinsights/datasets/coco.py [hunk 2]
```diff
@@ -173,3 +181,621 @@ def download(self, cloud_path=COCO_GCS_PATH):
...
+    def __init__(self, anchors, anchor_target, config, transforms):
+        global logger
+        logger = logging.getLogger("global")
+
+        self.split = "train"
+        # self.download()
+        # self.crop()
+        # self.generate_json_coco()
+        # All these hard-coded rignt now.
+        # TODO: Move these to config files
```
Comment (zookae) [inline datasetinsights/datasets/coco.py]: stray TODO

File: datasetinsights/datasets/coco.py [hunk 3]
```diff
@@ -173,3 +181,621 @@ def download(self, cloud_path=COCO_GCS_PATH):
...
+        self.anchors = anchors
+        self.template_size = 127
+        self.origin_size = 127
+        self.search_size = 255
+        # DO NOT change. TODO: Write tests for this
```
Comment (zookae) [inline datasetinsights/datasets/coco.py:199]: stray TODO

File: datasetinsights/datasets/coco.py [hunk 4]
```diff
@@ -173,3 +181,621 @@ def download(self, cloud_path=COCO_GCS_PATH):
...
+                    coco.loadAnns(coco.getAnnIds(imgIds=i, iscrowd=None)),
+                    set_crop_base_path,
+                    set_img_base_path,
+                    exemplar_size,
+                    context_amount,
+                    search_size,
+                    enable_mask,
+                )
+
+            # TODO: Figure out why async process pool is not working.
```
Comment (zookae) [inline datasetinsights/datasets/coco.py:467]: stray TODO

File: datasetinsights/datasets/coco.py [hunk 5]
```diff
@@ -173,3 +181,621 @@ def download(self, cloud_path=COCO_GCS_PATH):
...
+            #                           enable_mask) for id in coco.imgs]
+
+            #     for i, f in enumerate(futures.as_completed(fs)):
+            #         printProgress(i, n_imgs,
+            #         prefix=data_subset, suffix='Done ', barLength=40)
+
+    # Apply all transforms and return object
+    def __getitem__(self, index):
+        index = self.pick[index]
+        # TODO: put both negs in config files
```
Comment (zookae) [inline datasetinsights/datasets/coco.py:486]: stray TODO

File: datasetinsights/datasets/coco.py [hunk 6]
```diff
@@ -173,3 +181,621 @@ def download(self, cloud_path=COCO_GCS_PATH):
...
+            self.anchors, bbox, self.size, neg
+        )
+
+        # Change this if we want to train using any dataset
+        # which has no masks like vid
+
+        mask_weight = cls.max(axis=0, keepdims=True)
+        # if dataset.has_mask and not neg:
+        # else:
+        # mask_weight=np.zeros([1,cls.shape[1],cls.shape[2]],dtype=np.float32)
```
Comment (zookae) [inline datasetinsights/datasets/coco.py:557]: comment

File: datasetinsights/datasets/coco.py [hunk 7]
```diff
@@ -173,3 +181,621 @@ def download(self, cloud_path=COCO_GCS_PATH):
...
+        nsize = map_size(self.template_size, img.shape[1])
+
+        img = cv2.resize(img, (nsize, nsize))
+
+        return img, nsize / img.shape[1]
+
+    """
+        Populate function is used to read the dataset and load from the
+        json file created
+    """
```
Comment (zookae) [inline datasetinsights/datasets/coco.py:601]: should be moved inside function

File: datasetinsights/datasets/coco.py [hunk 8]
```diff
@@ -173,3 +181,621 @@ def download(self, cloud_path=COCO_GCS_PATH):
...
+                if len(self.labels[video]) <= 0:
+                    logger.info("warning {} has no tracks".format(video))
+                    to_del.append(video)
+
+            for video in to_del:
+                del self.labels[video]
+
+            self.videos = list(self.labels.keys())
+            # print("Total vids:",len(self.videos))
+            # exit(0)
```
Comment (zookae) [inline datasetinsights/datasets/coco.py:651]:
```suggestion
```

File: datasetinsights/datasets/coco.py [hunk 9]
```diff
@@ -173,3 +181,621 @@ def download(self, cloud_path=COCO_GCS_PATH):
...
+        # Frame range is x +/- 100 range.
+        # If kept high, the object might disappear from frame
+        self.frame_range = 100
+        self.path_format = "{}.{}.{}.jpg"
+        self.mask_format = "{}.{}.m.png"
+        self.pick = []
+
+        self.num_use = int(self.num_use)
+
+        # self.has_mask = self.mark in ["coco", "ytb_vos"]
```
Comment (zookae) [inline datasetinsights/datasets/coco.py:669]:
```suggestion
```

File: datasetinsights/datasets/coco.py [hunk 10]
```diff
@@ -173,3 +181,621 @@ def download(self, cloud_path=COCO_GCS_PATH):
...
+        # Kept here for generic purpose. Whenever new class is made-change this
+        self.has_mask = True
+
+        """
+        'Pick' is a list of index mappings from real to shuffled indices
+        To get consistent results the random variable with fixed seed can be set
+        """
+        # self.shuffle()
+        # print(self.__dict__.keys())
```
Comment (zookae) [inline datasetinsights/datasets/coco.py:678]:
```suggestion
```

File: datasetinsights/datasets/coco.py [hunk 11]
```diff
@@ -173,3 +181,621 @@ def download(self, cloud_path=COCO_GCS_PATH):
...
+
+    """
+    Function to shuffle the dataset as a whole.
+    Pick is created here
+    'Pick' is a list of index mappings from real to shuffled indices
+    To get consistent results the random variable with fixed seed can be set
+    """
```
Comment (zookae) [inline datasetinsights/datasets/coco.py:685]: move into function

File: datasetinsights/datasets/coco.py [hunk 12]
```diff
@@ -173,3 +181,621 @@ def download(self, cloud_path=COCO_GCS_PATH):
...
+            random.Random().shuffle(lists)
+            pick += lists
+            m += self.num
+
+        self.pick = pick[: self.num_use]
+        return self.pick
+
+    """
+    Function to remove the frames with 0 w/h
+    """
```
Comment (zookae) [inline datasetinsights/datasets/coco.py:701]: move into function

File: datasetinsights/datasets/coco.py [hunk 13]
```diff
@@ -173,3 +181,621 @@ def download(self, cloud_path=COCO_GCS_PATH):
...
+
+    def filter_zero(self, anno, dset, cfg):
+        name = dset
+        # cfg.get('mark', '')
+        # print(len(anno))
```
Comment (zookae) [inline datasetinsights/datasets/coco.py:706]:
```suggestion
```

File: datasetinsights/io/transforms.py [hunk 14]
```diff
@@ -59,3 +61,144 @@ def __call__(self, sample):
...
+            1 + (self.helper["random_number"](-1, 1) * self.scale),
+            1 + (self.helper["random_number"](-1, 1) * self.scale),
+        )
+        x1, y1, w, h = bbox.x, bbox.y, bbox.w, bbox.h
+        scale_x, scale_y = scale_val[0], scale_val[1]
+        # corner form and center form created
+        # calculate corner
+        # x2 = x1 + w
+        # y2 = y1 + h
+        # calculate center
```
Comment (zookae) [inline datasetinsights/io/transforms.py:195]: still needed?

File: datasetinsights/datasets/coco.py [hunk 15]
```diff
@@ -173,3 +181,621 @@ def download(self, cloud_path=COCO_GCS_PATH):
...
+            #                           iscrowd=None)),
+            #                           set_crop_base_path,
+            #                           set_img_base_path,
+            #                           exemplar_size,
+            #                           context_amount, search_size,
+            #                           enable_mask) for id in coco.imgs]
+
+            #     for i, f in enumerate(futures.as_completed(fs)):
+            #         printProgress(i, n_imgs,
+            #         prefix=data_subset, suffix='Done ', barLength=40)
```
Comment (zookae) [inline datasetinsights/datasets/coco.py:481]: stray TODO

File: datasetinsights/datasets/coco.py [hunk 16]
```diff
@@ -173,3 +181,621 @@ def download(self, cloud_path=COCO_GCS_PATH):
...
+                    > 0.5
+                ).astype(np.uint8) * 255
+                cv2.imwrite(
+                    join(
+                        frame_crop_base_path,
+                        "{:06d}.{:02d}.m.png".format(0, track_id),
+                    ),
+                    x,
+                )
+                # print("Done: ",'{:06d}.{:02d}.m.png'.format(0, track_id))
```
Comment (zookae) [inline datasetinsights/datasets/coco.py:375]:
```suggestion
```

**Comment by adason** [comment]
Since dataseingisht will not store model training code, I am closing this issue to reduce confusion.

**Pull request closed without merging**